### ISSUE TITLE
Transformers library owlv2

### ISSUE THREAD

**Issue by bw4sz** [weecology/DeepForest#915]
What would it take to get transformers as a dependency to DeepForest

https://huggingface.co/docs/transformers/main/en/model_doc/owlv2

seems interesting for open set plus learning.

**Comment by bw4sz** [comment]
# Roadmap

1. Add to requirements, conda and setup dependencies.
2. Add a simple transformers model to https://github.com/weecology/DeepForest/tree/main/src/deepforest/models
3. Add a attribute to deepforest.main to know whether we are torchvision versus transformers data format, looks like documented here https://huggingface.co/docs/transformers/tasks/object_detection
4. Update dataset.TreeDataset (https://github.com/weecology/DeepForest/blob/95c04d546d05be4a7eba0113e2b64525d3880243/src/deepforest/dataset.py#L48) to 1) yield the correct data format for torchvision versus transformers, 2) Flexible across geometries. This was a pretty old function and could be refactored. See https://github.com/weecology/DeepForest/blob/95c04d546d05be4a7eba0113e2b64525d3880243/src/deepforest/utilities.py#L251 for utility in reading geometry.
6. predict_step needs be flexible to geometry and torchvision versus transformers output https://github.com/weecology/DeepForest/blob/95c04d546d05be4a7eba0113e2b64525d3880243/src/deepforest/main.py#L852 
7. Write tests to check predict_image, predict_tile that covers both torchvision and transformers model.
8. Document transformers specific logic, which models can be used? How constant is their input architecture?

**Comment by naxatra2** [comment]
Hi @bw4sz ,
I read about this issue because it seems interesting. After reading the given resources from hugging face. I have this doubt:

Are we going to add wildlife detection models using transformers or tree crown detection. Because from what I read, Owl-ViT v2 is a general-purpose open-vocabulary detector, which is not specific for training that DeepForest offers, which is specialized for detecting tree crowns. Won't this make it less effective when it comes to recognizing the unique characteristics of tree crowns in forestry imagery? because transformers like this are not well-optimized for detecting overlapping tree crowns. This inefficiency is particularly problematic in the context of forestry, where canopy overlap detection is crucial. 

In [deepforest/models](https://github.com/weecology/DeepForest/tree/main/src/deepforest/models) there is already  Faster R-CNN model, so ontegrating Owl-ViT v2 into a pipeline that already uses this, could introduce unnecessary computational overhead without yielding significant improvements in accuracy.


But I still tried to create a simple file for the model, are we supposed to implement something like this. This file should go in `deepforest/models` directory as `owl.py`.
```py
from transformers import Owlv2Processor, Owlv2ForObjectDetection
from PIL import Image
import torch
from deepforest.model import Model

class OwlV2Model(Model):

    def __init__(self, config, **kwargs):
        super().__init__(config)

    def load_model(self):
        """Load Owl-ViT v2 for object detection."""
        processor = Owlv2Processor.from_pretrained("google/owlv2-base-patch16-ensemble")
        model = Owlv2ForObjectDetection.from_pretrained("google/owlv2-base-patch16-ensemble")
        return processor, model

    def detect_objects(self, image_path, texts):
        processor, model = self.load_model()

        image = Image.open(image_path).convert("RGB")

        inputs = processor(text=texts, images=image, return_tensors="pt")

        with torch.no_grad():
            outputs = model(**inputs)

        # Extract the predicted bounding boxes and scores
        logits = outputs.logits  # Prediction scores
        pred_boxes = outputs.pred_boxes  # Bounding boxes

        # Process and print results
        results = []
        for i, text in enumerate(texts):
            result_text = f"Query: {text}"
            for box, score in zip(pred_boxes[0], logits[0]):
                if score > 0.5:  # Confidence threshold
                    result_text += f"\n  - Detected {text} at {box.tolist()} with confidence {score.item():.2f}"
            results.append(result_text)
        
        return results
```

I am new to this topic and was reading from the hugging face docs, so if there are any improvement or any other useful resource please suggest something, then I will start working on this

**Comment by jveitchmichaelis** [comment]
Few things that came up with DETR:

- MillionTrees can be the data source of choice, DeepForest should be able handle the different formats required. This usually boils down to xyxy vs xywh for boxes. Polygons I would say standardise on COCO.
- I would place the Dataloader here and Dataset in MillionTrees. Models determine `preprocess` and thus data collation.
- I will make a tutorial script/doc that explains the differences in data formats and nuances in making transformers (the library) work.
- Will double check what I have in my DETR training loop, like vizualisation and other metrics that we don't have here.
- I would like to re-train the default retinanet model on MillionTrees and benchmark as a comparison to make sure the same metrics are used. Let's start the leaderboard!
- I chose Deformable DETR, would be worth checking Conditional DETR and the other variants as well but that might be a nice low-hanging fruit for a contributor.
- Other issues I'll flag in MillionTrees


Thoughts on Polygon and KeyPoint models? The obvious one is the *Former family (Mask2Former, OneFormer, etc), but I have no preference. When I tried these with Restor's data I had a lot of issues with CPU memory usage. Many polygons breaks the Hungarian matcher implementation, but I have ideas...

KeyPoints I have no idea except to integrate and retrain Treeformer on the "bulk" dataset from MillionTrees. Or we use a model like CountGD.

**Comment by jveitchmichaelis** [comment]
Yeah my thinking was that it would be helpful to have a comparison with retinanet so that we can quantify the difference between the different architectures. So for reproducibility and sanity checking than anything else. Especially since the current default DeepForest model was trained on only(?) NeonTreeEvaluation.
